File: banking-system/src/main/java/com/bank/banking_system/service/AccountService.java
package com.bank.banking_system.service;

import com.bank.banking_system.entity.Account;
import com.bank.banking_system.entity.AccountType;
import com.bank.banking_system.entity.Admin;
import com.bank.banking_system.entity.Branch;
import com.bank.banking_system.entity.Customer;
import com.bank.banking_system.exception.AccountNotFoundException;

import com.bank.banking_system.repository.AccountRepository;
import com.bank.banking_system.repository.AdminRepository;
import com.bank.banking_system.repository.BranchRepository;
import com.bank.banking_system.repository.CustomerRepository;
import com.bank.banking_system.repository.TransactionRepository;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

@Service
@Transactional
public class AccountService {

    private final AccountRepository accountRepository;
    private final CustomerRepository customerRepository;
    private final BranchRepository branchRepository;
    private final AdminRepository adminRepository;
    private final TransactionRepository transactionRepository;

    public AccountService(AccountRepository accountRepository,
                          CustomerRepository customerRepository,
                          BranchRepository branchRepository,
                          AdminRepository adminRepository,
                          TransactionRepository transactionRepository) {

        this.accountRepository = accountRepository;
        this.customerRepository = customerRepository;
        this.branchRepository = branchRepository;
        this.adminRepository = adminRepository;
        this.transactionRepository = transactionRepository;
    }

    // ===== OPEN ACCOUNT =====
    public Account openAccount(Long customerId,
                               Long branchId,
                               Long adminId,
                               String accountNumber,
                               AccountType type,
                               BigDecimal balance) {

        Customer customer = customerRepository.findById(customerId)
                .orElseThrow(() -> new RuntimeException("Customer not found"));

        Branch branch = branchRepository.findById(branchId)
                .orElseThrow(() -> new RuntimeException("Branch not found"));

        Admin admin = adminRepository.findById(adminId)
                .orElseThrow(() -> new RuntimeException("Admin not found"));

        Account account = new Account();
        account.setAccountNumber(accountNumber);
        account.setAccountType(type);
        account.setBalance(balance);
        account.setCustomer(customer);
        account.setBranch(branch);
        account.setAdmin(admin);
        account.setOpenedDate(LocalDate.now());
        account.setStatus("ACTIVE");

        return accountRepository.save(account);
    }

    // ===== GET ALL ACCOUNTS =====
    public List<Account> getAllAccounts() {
        return accountRepository.findAll();
    }

    // ===== GET ACCOUNT BY ID =====
    public Account getAccountById(Integer accountId) {
        return accountRepository.findById(accountId)
                .orElseThrow(() -> new RuntimeException("Account not found"));
    }

    // ===== GET BALANCE =====
    public BigDecimal getBalance(Integer accountId) {

        Account account = accountRepository.findById(accountId)
                .orElseThrow(() -> new RuntimeException("Account not found"));

        return account.getBalance();
    }

    // ===== UPDATE ACCOUNT =====
    public Account updateAccount(Integer accountId, AccountType type, BigDecimal balance, String status) {
        Account account = accountRepository.findById(accountId)
                .orElseThrow(() -> new RuntimeException("Account not found"));

        if (type != null) account.setAccountType(type);
        if (balance != null) account.setBalance(balance);
        if (status != null) account.setStatus(status);

        return accountRepository.save(account);
    }

    // ===== DELETE ACCOUNT =====
    public void deleteAccount(Integer accountId) {
        Account account = accountRepository.findById(accountId)
                .orElseThrow(() -> new RuntimeException("Account not found"));
        
        // Delete all transactions associated with this account first
        transactionRepository.deleteByAccount_AccountId(accountId);
        
        // Now delete the account
        accountRepository.delete(account);
    }

    // ===== GET ACCOUNTS BY CUSTOMER ID =====
    public List<Account> getAccountsByCustomerId(Long customerId) {
        return accountRepository.findByCustomer_CustomerId(customerId);
    }

    // ===== GET ACCOUNT BY ACCOUNT NUMBER =====
    public Account getAccountByNumber(String accountNumber) {
        // Try exact match first, then case-insensitive
        Account account = accountRepository.findByAccountNumber(accountNumber);
        if (account == null) {
            account = accountRepository.findByAccountNumberIgnoreCase(accountNumber);
        }
        if (account == null) {
            throw new AccountNotFoundException("Account not found with number: " + accountNumber);
        }
        return account;
    }
}

File: banking-frontend/src/app/services/account.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Account } from '../models/account.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class AccountService {

  // ✅ ADMIN BASE API
  private adminUrl = `${environment.apiUrl}/api/admin`;

  constructor(private http: HttpClient) {}

  // ================= GET =================

  getAllAccounts(): Observable<Account[]> {
    return this.http.get<Account[]>(`${this.adminUrl}/accounts`);
  }

  getAccountById(id: number): Observable<Account> {
    return this.http.get<Account>(`${this.adminUrl}/account/${id}`);
  }

  getAccountsByCustomerId(customerId: number): Observable<Account[]> {
    return this.http.get<Account[]>(`${environment.apiUrl}/api/customer/accounts/${customerId}`);
  }

  // ================= CREATE =================

  createAccount(data: any): Observable<Account> {
    const params = {
      customerId: data.customerId,
      branchId: data.branchId || 1,  // Default branch
      adminId: data.adminId || 1,    // Default admin
      accountNumber: data.accountNumber,
      type: data.accountType,
      balance: data.balance
    };
    return this.http.post<Account>(`${this.adminUrl}/open-account`, null, { params });
  }

  // ================= UPDATE =================

  updateAccount(id: number, data: any): Observable<Account> {
    const params: any = {};
    if (data.accountType) params.type = data.accountType;
    if (data.balance !== undefined) params.balance = data.balance;
    if (data.status) params.status = data.status;
    return this.http.put<Account>(`${this.adminUrl}/account/${id}`, null, { params });
  }

  // ================= DELETE =================
  // (Only works if backend endpoint exists)

  deleteAccount(id: number): Observable<void> {
    return this.http.delete<void>(`${this.adminUrl}/account/${id}`);
  }

  // ================= CUSTOMER ACTIONS =================

  deposit(accountId: number, amount: number): Observable<any> {
    return this.http.post(
      `${environment.apiUrl}/api/customer/deposit?accountId=${accountId}&amount=${amount}`,
      {}
    );
  }

  withdraw(accountId: number, amount: number): Observable<any> {
    return this.http.post(
      `${environment.apiUrl}/api/customer/withdraw?accountId=${accountId}&amount=${amount}`,
      {}
    );
  }

}



